Use autolinking to pull in compatibility libraries.

jckarter opened the pull request:
Many build systems that support Swift don't use swiftc to drive the linker. To make things
easier for these build systems, also use autolinking to pull in the needed compatibility
libraries. This is less ideal than letting the driver add it at link time, since individual
compile jobs don't know whether they're building an executable or not. Introduce a
`-disable-autolink-runtime-compatibility` flag, which build systems that do drive the linker
with swiftc can pass to avoid autolinking.

rdar://problem/50057445

jckarter committed:
Use autolinking to pull in compatibility libraries.

Many build systems that support Swift don't use swiftc to drive the linker. To make things
easier for these build systems, also use autolinking to pull in the needed compatibility
libraries. This is less ideal than letting the driver add it at link time, since individual
compile jobs don't know whether they're building an executable or not. Introduce a
`-disable-autolink-runtime-compatibility` flag, which build systems that do drive the linker
with swiftc can pass to avoid autolinking.

rdar://problem/50057445

jckarter commented:
@swift-ci Please test

jrose-apple commented on lib/Basic/Version.cpp:
```diff
@@ -444,6 +445,41 @@ std::string getSwiftRevision() {
 }
+  
+Optional<llvm::VersionTuple>
+getSwiftRuntimeCompatibilityVersionForTarget(const llvm::Triple &Triple) {
```
Can you move this to Platform.cpp, which is doing existing checks like this?

jrose-apple commented on stdlib/public/Compatibility50/Overrides.cpp:
```diff
@@ -31,3 +31,8 @@ __attribute__((used, section("__DATA,__swift_hooks"))) = {
 };
+
+// Allow this library to get force-loaded by autolinking
+__attribute__((weak, visibility("hidden")))
```
I don't think you can make this weak (Apple's build system will complain about it). I believe actual IRGen uses a function so it can be linkonce_odr.

jckarter replied:
`weak` is only a problem for symbols with default visibility.

jrose-apple replied:
Oh, right, good point. Withdrawn.

jrose-apple reviewed (commented):
A few comments. Also needs a test case.

jckarter commented:
@swift-ci Please test

jckarter commented:
@jrose-apple How's this look now?

jckarter commented:
@swift-ci Please test

jrose-apple commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -431,12 +396,13 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
   if (context.Args.hasArg(options::OPT_runtime_compatibility_version)) {
-    auto value = context.Args.getLastArgValue(options::OPT_runtime_compatibility_version);
+    auto value = context.Args.getLastArgValue(
+                                    options::OPT_runtime_compatibility_version);
```
Nitpick (here and elsewhere): double-indent for wrapped lines rather than reverse-indenting until it fits.

compnerd replied:
`clang-format` should DTRT

jrose-apple commented on test/IRGen/autolink-runtime-compatibility.swift:
```diff
@@ -0,0 +1,26 @@
+
+// NO-FORCE-LOAD-NOT: FORCE_LOAD
+// NO-FORCE-LOAD-NOT: !{!"-lswiftCompatibility50"}
+// NO-FORCE-LOAD-NOT: FORCE_LOAD
```
You don't need to repeat this check because the entire `NO-FORCE-LOAD` check is negative.

jrose-apple commented on test/IRGen/autolink-runtime-compatibility.swift:
```diff
@@ -0,0 +1,26 @@
+// FORCE-LOAD: declare {{.*}} @"_swift_FORCE_LOAD_$_swiftCompatibility50"
+
+// FORCE-LOAD-DAG: [[AUTOLINK_SWIFT_COMPAT:![0-9]+]] = !{!"-lswiftCompatibility50"}
+// FORCE-LOAD-DAG: !llvm.linker.options = !{{{.*}}[[AUTOLINK_SWIFT_COMPAT]]{{[,}]}}
```
I'm surprised this works! Since it's matching in reverse order from what it needs. But if it works then okay. (If it doesn't actually work, and it's worth checking, then I think it's okay to just match the first line and trust that we wouldn't generate that for any other reason.)

jckarter replied:
I thought that was what `-DAG` was for, to allow lines to be matched out of order from how they show up in the output.

jrose-apple replied:
It is, but I've never tried using that with captures.

jrose-apple reviewed (approved):
Looks good, with only a few remaining minor comments. Make sure you're coordinating with @Rostepher about how this will get installed into the toolchain.

jckarter commented:
@swift-ci Please test

jckarter committed:
Install swiftCompatibility50 as part of the compiler component

jckarter commented:
@swift-ci Please test

compnerd commented on lib/Basic/Platform.cpp:
```diff
@@ -313,3 +314,37 @@ llvm::Triple swift::getTargetSpecificModuleTriple(const llvm::Triple &triple) {
+  
+  if (Triple.isMacOSX()) {
+    Triple.getMacOSXVersion(Major, Minor, Micro);
+    if (Major == 10) {
```
I think that you can write this more concisely as:

```
  if (Major == 10 && Minor <= 14)
    return llvm::VersionTuple(5, 0);
  return None;
```

jckarter replied:
I'm writing this with an eye toward it eventually growing more arms. There might be more macOS versions with Swift in them in the near future, but who knows

compnerd commented on lib/Basic/Platform.cpp:
```diff
@@ -313,3 +314,37 @@ llvm::Triple swift::getTargetSpecificModuleTriple(const llvm::Triple &triple) {
+    }
+  } else if (Triple.isiOS()) { // includes tvOS
+    Triple.getiOSVersion(Major, Minor, Micro);
+    if (Major <= 12) {
```
No need for the `else` after `return`

compnerd commented on lib/Basic/Platform.cpp:
```diff
@@ -313,3 +314,37 @@ llvm::Triple swift::getTargetSpecificModuleTriple(const llvm::Triple &triple) {
+    Triple.getWatchOSVersion(Major, Minor, Micro);
+    if (Major <= 5) {
+      return llvm::VersionTuple(5, 0);
+    } else {
```
No need for the `else` after `return`

compnerd reviewed (commented):
Ugh, I really should finish up my work to introduce the SwiftRuntime thing ... this is becoming a mess (not your change, just all the places we have these adhoc checks)

jckarter committed:
Install compatibility archive alongside shared libraries.

Although it's a static archive, its use is only relevant to dynamically linked builds of the
standard library. (If you're statically linking a Swift runtime into your process, you don't
need to worry about compatibility with older runtimes.)

jckarter commented:
@swift-ci Please test

Rostepher commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1085,18 +1086,24 @@ function(_add_swift_library_single target name)
         BINARY_DIR ${SWIFT_RUNTIME_OUTPUT_INTDIR}
         LIBRARY_DIR ${SWIFT_LIBRARY_OUTPUT_INTDIR})
 
+    if(SWIFTLIB_INSTALL_WITH_SHARED)
```
For this sort of logic I usually prefer:

```CMake
set(swift_lib_dir "${SWIFTSTATICLIB_DIR}")
if(SWIFTLIB_INSTALL_WITH_SHARED)
  set(swift_lib_dir "${SWIFTLIB_DIR}")
endif()
```

Rostepher commented on cmake/modules/AddSwift.cmake:
```diff
@@ -730,7 +730,8 @@ function(_add_swift_library_single target name)
         STATIC
-        TARGET_LIBRARY)
+        TARGET_LIBRARY
+        INSTALL_WITH_SHARED)
```
I'm not sure I agree with adding yet another single purpose flag to `add_swift_target_library`. Is there a more generic purpose here? Say maybe an `INSTALL_PATH` or something?

jckarter replied:
Yeah, I'm torn about that too. YAGNI convinced me to stick with just the flag for now. A small generalization in between this and INSTALL_PATH might be to have an `INSTALL_AS [SHARED|STATIC]` flag, so you could also choose to send dylibs to the static dir and v.v.?
